{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "changed-centre",
   "metadata": {
    "_kg_hide-input": true,
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:24.351815Z",
     "iopub.status.busy": "2021-09-08T13:53:24.349996Z",
     "iopub.status.idle": "2021-09-08T13:53:24.367219Z",
     "shell.execute_reply": "2021-09-08T13:53:24.366555Z"
    },
    "papermill": {
     "duration": 0.048513,
     "end_time": "2021-09-08T13:53:24.367388",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.318875",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from re import sub, search\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import math\n",
    "\n",
    "''' Display List '''\n",
    "# display list neatly\n",
    "# https://stackoverflow.com/questions/1524126/how-to-print-a-list-more-nicely\n",
    "def lstcol(obj, cols=4, columnwise=True, gap=4):\n",
    "    sobj = [str(item) for item in obj]\n",
    "    if cols > len(sobj): cols = len(sobj)\n",
    "    max_len = max([len(item) for item in sobj])\n",
    "    if columnwise: cols = int(math.ceil(float(len(sobj)) / float(cols)))\n",
    "    plist = [sobj[i: i+cols] for i in range(0, len(sobj), cols)]\n",
    "    if columnwise:\n",
    "        if not len(plist[-1]) == cols:\n",
    "            plist[-1].extend(['']*(len(sobj) - len(plist[-1])))\n",
    "        plist = zip(*plist)\n",
    "    printer ='\\n'.join([\n",
    "        ''.join([c.ljust(max_len + gap) for c in p])\n",
    "        for p in plist])\n",
    "    print (printer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "british-pledge",
   "metadata": {
    "papermill": {
     "duration": 0.0282,
     "end_time": "2021-09-08T13:53:24.420866",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.392666",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "![](https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/8cc1eeaa-4046-4c4a-ae93-93d656f68688/deny9ja-e5f8e0f2-94f0-402a-9350-32ebbf1fb61d.jpg?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcLzhjYzFlZWFhLTQwNDYtNGM0YS1hZTkzLTkzZDY1NmY2ODY4OFwvZGVueTlqYS1lNWY4ZTBmMi05NGYwLTQwMmEtOTM1MC0zMmViYmYxZmI2MWQuanBnIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.y0MmVL5Y2FuoS12AgjAJLAGOWtpd-WYLfzg7gfNu_1g)\n",
    "Photography of <b>Evolution</b> by [@sashayudaev](https://unsplash.com/@sashayudaev)\n",
    "\n",
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>1 |</b> BACKGROUND\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "### <b><span style='color:#FF5733'>1.1</span> | CELLS</b>\n",
    "- A cell is mostly composed of water, a bacteria cell has a weight composition of roughly 70% water & 30% chemical origin, of which <b>7% are small molecules</b>, inc. <b>amino acids</b> and <b>nucleotides</b> & 23% -> <b>macro molecules</b> (<b>proteins</b>,lipids,polysaccharides)\n",
    "- According to their internal structure, they can be divided into to major categories; \n",
    "  - <b>Prokaryotic</b> cells : have no nucleus or internal membranes. \n",
    "  - <b>Eukaryotic</b> cells : which have a defined <b>nucleus</b>, <b>internal membranes</b> and functional elements called <b>organelles</b>; have varying shapes and sizes and play specific functions.\n",
    "- At a structural level, all cells are surrounded by a structure called cell membrane or plasma membrane. This membrane is permeable to molecules that cells need to absorb from or excrete to the outside medium.\n",
    "- Within the cell we find the <b>cytoplasm</b> (largely composed of water), which serves as the medium for the cell.\n",
    "\n",
    "#### <b><span style='color:#FF5733'>NUCLEIC ACIDS</span></b>\n",
    "- Among molecules with a biological role, we can find <b>nucleic acids</b>. \n",
    "> Nucleic acids encode and express the genetic code that is kept within the cell. \n",
    "- There are two major types of nucleic acids: \n",
    "  - (a) <b>DeoxyriboNucleic Acid (DNA)</b>\n",
    "  - (b) <b>RiboNucleic Acid (RNA)</b>\n",
    "- DNA contains the information necessary to build a cell, and keep it functioning. \n",
    "- In <b>eukaryotic</b> cells, DNA will be found in the nucleus, whilst in the <b>prokaryotic</b> cells, it will be found in the cytoplasm. \n",
    "    \n",
    "|Name|Symbol|Nucleotides| Name | Symbol | Nucleotides |\n",
    "|-|-|-|-|-|-|\n",
    "|Adenine|A|A|Amino|M|A,C|\n",
    "|Cytosine|C|C|Keto|K|G,T|\n",
    "|Guanine|G|G|Purine|R|A,G|\n",
    "|Thymine|T|T|Pyrimidine|Y|C,T|\n",
    "|Uracil|U|U|Strong|S|C,G|\n",
    "|Strong|S|C,G|Weak|W|A,T|\n",
    "\n",
    "#### <b><span style='color:#FF5733'>AMINO ACIDS</span></b>\n",
    "- Amino acids: The <b>building blocks of proteins</b>, which are <b>macromolecules</b> that perform most of the functions in a cell.\n",
    "- Proteins have a broad range of functions, spanning from catalytic to structural functions:\n",
    "  - <b>Enzymes</b> : type of abundant proteins that promote chemical reactions and convert certain molecules into other types of molecules required for the functioning of the cell.\n",
    "  - <b>Carbohydrates</b> : serve as energy storage, both for immediate and long term energy demands.\n",
    "  - <b>Lipids</b>, which are part of the plasma membrane, doing signaling and energy storage.\n",
    "- The cell also contains other components of varying complexity. Of importance: \n",
    "  - <b>mitochondria</b> & the <b>chloroplasts</b> : organelles involved in the production of energy. \n",
    "  - <b>Ribosomes</b> : Large and complex molecules composed of a mixture of genetic material, required to assemble proteins and play a central role in the flow of genetic information.\n",
    "- Some Aminoacids (as per IUPAC amino acid list):\n",
    "    \n",
    "| Name | Symbol | Name | Symbol | Name | Symbol | \n",
    "|-|-|-|-|-|-|\n",
    "| Alanine | A | Arginine | R | Aspartic acid | D |\n",
    "| Asparagine | N | Cysteine | C | Glycine | G |\n",
    "| Glutamic Acid | E | Glutamine | Q | Histidine | H | \n",
    "| Isoleucine | I | Leucine | L | Lysine | K |\n",
    "| Methionine | M | Proline | P | Phenylalanine | F | \n",
    "| Serine | S | Threonine | T | Tryptophan | W | \n",
    "| Tyrosine | Y | Valine | V | \n",
    "\n",
    "### <b><span style='color:#FF5733'>1.2</span> | NUCLEIC ACID</b>\n",
    "- DNA is a molecule composed of <b>two complementary strands</b> that form and stick together due to the connections established between the nucleotides in both strands. \n",
    "- This is made possible by due to the chemical phenomenon where <b>Adenine</b> (A), bonds only with <b>Thymine</b> (T) nucleotides, as a result of two hydrogen connections. Similarly, <b>Guanine</b> (G) bonds only with <b>Cytosine</b> (C) nucleotides by three hydrogen connections.\n",
    "- This results in <b>two complementary</b> and anti-parallel strands (connected in opposite directions), if we know the nucleotide sequence in one of the strands, we can get the sequence in the opposite strand by taking the complement of its nucleotides, which are also read backwards, thus we have the <b>reverse complement</b> of the other strand. It has become a standard to describe the DNA though only one of the strands, due to this complementarity using <b>[A,T,G,C]</b>.\n",
    "- The existence of these two strands is essential in order to pass on genetic information to new cells and produce <b>proteins</b>.\n",
    "\n",
    "### <b><span style='color:#FF5733'>1.3</span> | CENTRAL DOGMA OF MOLECULAR & CELL BIOLOGY</b>\n",
    "\n",
    "- <b>DNA</b>, <b>RNA</b> & <b>Proteins</b> are the central elements of the flow of genetic information that occurs in two steps (1) <b>transcription</b> & (2) <b>translation</b>.\n",
    "\n",
    "#### <b>(1) RNA Synthesis : <span style='color:#FF5733'>Transcription</span></b>\n",
    "\n",
    "- *Transcription* : Preliminary step required to produce a <b>protein</b>, as shown in Fig. X.\n",
    "> - The nucleotide sequence of a gene from one of the DNA strands is transcribed ( copied into a complementary molecule of RNA )\n",
    "> - The complementarity of the genetic code allows recovering the information encoded in the original DNA sequence, a process performed by the enzyme, RNA polymerase.\n",
    "> - Additional steps of RNA processing, including stabilising elements at the end of the molecule, are performed by different protein complexes.\n",
    "> - After these steps, which occur within the nucleus of the cell, an RNA molecule; <b>mature messenger RNA (mRNA)</b> is obtained.\n",
    "> - The <b>mRNA</b> is then transported to the cytoplasm, where it will be used by the cellular machine to guide the production of a protein.\n",
    "\n",
    "#### <b>(2) Protein Synthesis : <span style='color:#FF5733'>Translation</span></b>\n",
    "> - <b>Proteins</b> are cellular entities that have either:\n",
    "    - (a) <b>structural function</b>, participating in the physical definition of a cell.\n",
    "    - (b) <b>chemical function</b>; being involved in chemical reactions occuring in the cell.\n",
    "> -  In order to function as expected, a protein needs to acquire the appropriate structure & this structure is often decomposed at different complexity levels.\n",
    "> The primary structure is defined by the chain of amino acids; <b>polypeptide</b>, consisting in part or completely of protein.\n",
    "\n",
    "> - *Translation* - process in which the nucleotide sequence of the mRNA is <b>transcribed into a chain of amino acids</b>, forming a <b>polypeptide</b>. \n",
    "\n",
    "<b>What performs this?</b>\n",
    "> - This <b>process is performed by the ribosomes</b> that attach and scan the mRNA from one end to the other, in groups of <b>nucleotide triplets</b> / <b>Codons</b>.\n",
    "\n",
    "<b>Codons</b>\n",
    "    \n",
    "> - In each position of the triplet, we have 1/4 nucleotides, ie. there are 4x4x4 (64) possible triplets/Codons. \n",
    "> - For each codon in the mRNA sequence, we have a corresponding amino acid in the <b>polypeptide chain</b>.\n",
    "    \n",
    "<b>Start & Stop Codons</b>\n",
    "    \n",
    "> - Some of these codons represent significant signals that indicate the <b>initiation</b> or the <b>termination</b> of the translation process.\n",
    "> - Once the **ribosome** detects an initiation codon, it starts the formation of the amino acid chain, and when it scans the stop codon, it stops the translation and detaches from the mRNA molecule.\n",
    "\n",
    "> - There are <b>20 types of amino acids used to form polypeptides</b> (for IUPAC) & less than the 64 possible codons, therefore we have more than one codon corresponds to a type of amino acid.\n",
    "    \n",
    "> - During the translation process:\n",
    "    - A type of small RNA molecule, <b>transfer RNA (tRNAs)</b> will bring to the <b>ribosome</b>, the amino acids of the corresponding type, which will be complementary to the mRNA codon that is currently being scanned.\n",
    "    - Each mRNA molecule can be scanned multiple times by different ribosomes, giving rise to multiple copies of the polypeptide. \n",
    "    - With its redundency, where more than one codon encodes an amino acid, the genetic code encloses a very efficient code-correction mechanism that minimises the impact of errors in the nucleotide sequence occuring in DNA replication.\n",
    "\n",
    "**ORF**\n",
    "    \n",
    "> - During translation process:\n",
    "    - Parsing of the mRNA sequence by the ribosome may start at different nucleotides. \n",
    "    - Given that a codon is composed of three nucleotides, the mRNA sequence may have <b>3 possible interpretations</b>. \n",
    "    - These three ways of parsing the sequence are called; <b>reading frames</b>.\n",
    "\n",
    "### **<span style='color:#FF5733'>1.4</span> | CUSTOM SEQUENCE CLASS**\n",
    "- Whilst there are wonderful libraries like <b>BioPython</b>, it's often quite benefitial to understand the workings of the code & expand on the code if a need arises. Whilst this is more time consuming, it's definitely more interesting.\n",
    "- In this notebook, emphasis will be placed on replicating something similar to two BioPython modules.\n",
    "> from Bio import SeqIO  ( Class for readng sequences ) <br>\n",
    "> from Bio.Seq import Seq ( Class for Sequence Operations )\n",
    "\n",
    "### **<span style='color:#FF5733'>1.5</span> | USEFUL REFERENCE**\n",
    "- [BioPython GenBank](https://github.com/biopython/biopython/tree/master/Tests/GenBank) ( Small collection of files containing readable sequences )\n",
    "- [Sequence Similarity Searching](https://www.ebi.ac.uk/Tools/sss/) ( Sequence related database )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "internal-wallpaper",
   "metadata": {
    "papermill": {
     "duration": 0.027794,
     "end_time": "2021-09-08T13:53:24.479022",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.451228",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>2 |</b> WORKING WITH SEQUENCES\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "### **<span style='color:#FF5733'>2.1</span> | MAPPING DICTIONARIES**\n",
    "- Similar to BioPython's <b>CodonTable</b> as shown in [Biopython for Bioinformatics Basics](https://www.kaggle.com/shtrausslearning/biopython-for-bioinformatics-basics), a dictionary containing mapping data between different combinations of mRNA nucleotides (codons) and <b>amino acids</b> is obtainable in <b>dic_map</b> by setting <b>map_id</b> to *codon*. \n",
    "- Similarly, mapping between the IUPAC amino acid symbols & names can be obtained by using <b>map_id = 'iupac_amino'</b> &  <b>map_id = 'iupac_nucleotide'</b> respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "least-authentication",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:24.551695Z",
     "iopub.status.busy": "2021-09-08T13:53:24.550980Z",
     "iopub.status.idle": "2021-09-08T13:53:24.553059Z",
     "shell.execute_reply": "2021-09-08T13:53:24.553498Z"
    },
    "papermill": {
     "duration": 0.04427,
     "end_time": "2021-09-08T13:53:24.553678",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.509408",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mapping Dictionary\n",
    "def dic_map(map_id='codon',tid=None):\n",
    "    \n",
    "    # Codon / Amino Acid Conversion\n",
    "    if(map_id is 'codon'):\n",
    "        tc = {\"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",  # Alanine\n",
    "              \"TGT\":\"C\", \"TGC\":\"C\",    # Cystene\n",
    "              \"GAT\":\"D\", \"GAC\":\"D\",    # Aspartic Acid \n",
    "              \"GAA\":\"E\", \"GAG\":\"E\",    # Glutamic Acid\n",
    "              \"TTT\":\"F\", \"TTC\":\"F\",    # Phenylalanine\n",
    "              \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",  # Glutamic Acid\n",
    "              \"CAT\":\"H\", \"CAC\":\"H\",            # Histidine\n",
    "              \"ATA\":\"I\", \"ATT\":\"I\", \"ATC\":\"I\",   # Isoleucine\n",
    "              \"AAA\":\"K\", \"AAG\":\"K\",              # Lycine\n",
    "              \"TTA\":\"L\", \"TTG\":\"L\", \"CTT\":\"L\",   # Leucine\n",
    "              \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",   # Leucine\n",
    "              \"ATG\":\"M\",                         # Start Codon | Methionine \n",
    "              \"AAT\":\"N\", \"AAC\":\"N\",             # Asparagine\n",
    "              \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",  # Proline\n",
    "              \"CAA\":\"Q\", \"CAG\":\"Q\",              # Glutamine\n",
    "              \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\",   # Arginine\n",
    "              \"CGG\":\"R\", \"AGA\":\"R\", \"AGG\":\"R\",   # Arginine\n",
    "              \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\",   # Serine\n",
    "              \"TCG\":\"S\", \"AGT\":\"S\", \"AGC\":\"S\",   # Serie\n",
    "              \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",  # Threonine\n",
    "              \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",  # Valine\n",
    "              \"TGG\":\"W\",                        # Tryptophan\n",
    "              \"TAT\":\"Y\", \"TAC\":\"Y\",             # Tryosine\n",
    "              \"TAA\":\"_\", \"TAG\":\"_\", \"TGA\":\"_\" # Stop Codons | Gap\n",
    "            }\n",
    "    \n",
    "    # IUPAC Amino Acids\n",
    "    elif(map_id is 'iupac_amino'):\n",
    "        tc   = {'A':'Alanine','C':'Cysteine','D':'Aspartic Acid','E':'Glutamic Acid',\n",
    "                'F':'Phenylalanine','G':'Glycine','H':'Histidine','I':'Isoleucine',\n",
    "                'L':'Lysine','M':'Methionine','N':'Asparagine','P':'Proline',\n",
    "                'Q':'Glutamine','R':'Arginine','S':'Serine','T':'Threonine',\n",
    "                'V':'Valine','W':'Tryptophan','Y':'Tryosine','_':'Gap'}\n",
    "       \n",
    "    # IUPAC nuceotides\n",
    "    elif(map_id is 'iupac_nucleotide'):\n",
    "        tc  = {'A':'Adenine','C':'Cytosine',\n",
    "               'G':'Guanine','T':'Thymine',\n",
    "               'U':'Uracil'}\n",
    "    \n",
    "    if tid in tc: \n",
    "      return tc[tid]\n",
    "    else: \n",
    "      return None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lasting-quarter",
   "metadata": {
    "papermill": {
     "duration": 0.024776,
     "end_time": "2021-09-08T13:53:24.603649",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.578873",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>2.2</span> | MAIN SEQUENCE CLASS, SQ()**\n",
    "- We define a custom class, <b>SQ</b>, which stores <b>information about the sequence</b> & contains basic sequence related operations.\n",
    "- The <b>sequence class (SQ)</b> can be coupled with other classes that incorporate more sophisticated sequence based operations; eg. <b>sequence alignment</b>, and is passed on as a reference to other classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "happy-parent",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:24.658337Z",
     "iopub.status.busy": "2021-09-08T13:53:24.657491Z",
     "iopub.status.idle": "2021-09-08T13:53:26.124192Z",
     "shell.execute_reply": "2021-09-08T13:53:26.123447Z"
    },
    "papermill": {
     "duration": 1.495656,
     "end_time": "2021-09-08T13:53:26.124354",
     "exception": false,
     "start_time": "2021-09-08T13:53:24.628698",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from collections import Counter\n",
    "\n",
    "def dict_sum(dictlist):\n",
    "  outdic = {}\n",
    "  for d in dictlist:\n",
    "    for k in d.keys():\n",
    "      outdic[k] = 0\n",
    "  for d in dictlist:\n",
    "    for k in d.keys():\n",
    "      outdic[k]+=d[k]\n",
    "  return outdic\n",
    "\n",
    "# Class for Sequence Operations \n",
    "class SQ: \n",
    "    \n",
    "    '''Constructor'''\n",
    "    def __init__ (self, seq=None, seq_type = \"DNA\"): \n",
    "        self.seq = seq.upper()\n",
    "        self.seq_type = seq_type\n",
    "          \n",
    "    # class instance operations\n",
    "    def __len__(self):\n",
    "        return len(self.seq)\n",
    "    def __getitem__(self, n):\n",
    "        return self.seq[n]\n",
    "    def __getslice__(self, i, j):\n",
    "        return self.seq[i:j]\n",
    "    def __str__(self):\n",
    "        return self.seq\n",
    "    \n",
    "    '''Get frequency of all sequence symbols'''\n",
    "    def freq(self,compare=None,show_id='perc',fheight=None,fwidth=None):\n",
    "        \n",
    "        if(compare is not None):\n",
    "            if(self.seq_type != compare.seq_type):\n",
    "                print('sequences are not of same type')\n",
    "                return None\n",
    "            \n",
    "        c1 = dict(Counter(self.seq))  # abc counter for s1\n",
    "        if(compare is not None):\n",
    "            c2 = dict(Counter(compare))  # abc counter for s2\n",
    "            \n",
    "        abc = list(self.abc())\n",
    "        count = Counter(abc)\n",
    "        abc_c = dict(Counter({x:0 for x in count}))\n",
    "        \n",
    "        c_all1 = dict_sum([c1,abc_c])\n",
    "        if(compare is not None):\n",
    "            c_all2 = dict_sum([c2,abc_c])    \n",
    "\n",
    "        lst = []\n",
    "        for i in c_all1.keys():\n",
    "           if(self.seq_type == 'DNA' or self.seq_type == 'mRNA'):\n",
    "               lst.append(dic_map('iupac_nucleotide',i))\n",
    "           elif(self.seq_type == 'PROTEIN'):\n",
    "               lst.append(dic_map('iupac_amino',i))\n",
    "                \n",
    "        if(compare is not None):\n",
    "            lst2 = []\n",
    "            for i in c_all2.keys():\n",
    "               if(self.seq_type == 'DNA' or self.seq_type == 'mRNA'):\n",
    "                   lst2.append(dic_map('iupac_nucleotide',i))\n",
    "               elif(self.seq_type == 'PROTEIN'):\n",
    "                   lst2.append(dic_map('iupac_amino',i))\n",
    "          \n",
    "        perc = [round(x / len(self.seq),3) for x in [*c_all1.values()]]\n",
    "        if(show_id is 'perc'):\n",
    "            show1 = lst; show2 = perc\n",
    "        elif(show_id is 'count'):\n",
    "            show1 = lst; show2 = [*c_all1.values()]\n",
    "        fig = go.Figure(go.Bar(y=show1,x=show2,\n",
    "                               marker_color='rgb(26, 118, 255)',\n",
    "                               orientation='h',text=show2,\n",
    "                               textposition='outside',name='SEQ1'))\n",
    "        if(compare is not None):\n",
    "            perc = [round(x / len(compare),3) for x in [*c_all2.values()]]\n",
    "            if(show_id is 'perc'):\n",
    "                show1 = lst2; show2 = perc\n",
    "            elif(show_id is 'count'):\n",
    "                show1 = lst2; show2 = [*c_all2.values()]\n",
    "            fig.add_trace(go.Bar(y=show1,x=show2,marker_color='rgb(55, 83, 109)',\n",
    "                                 orientation='h',text=show2,\n",
    "                                 textposition='outside',name='SEQ2'))\n",
    "        fig.update_layout(template='plotly_white',height=fheight,width=fwidth,\n",
    "                         title=f'<b>{self.seq_type} SEQUENCE CONTENT</b>',\n",
    "                         font=dict(family='sans-serif',size=12),\n",
    "                         margin=dict(l=40, r=40, t=50, b=10))\n",
    "        fig.show()\n",
    "\n",
    "    '''Return percentage of G & C Nucleotides in sequence'''\n",
    "    def gc(self):\n",
    "        if (self.seq_type == \"DNA\" or self.seq_type == \"mRNA\"):\n",
    "            ii = 0\n",
    "            for s in self.seq:\n",
    "                if(s in \"GCgc\"):\n",
    "                    ii += 1\n",
    "            return round(ii / len(self.seq),4)\n",
    "        else:\n",
    "            return None\n",
    "        \n",
    "    '''General sequence information'''\n",
    "    def get_seq_biotype (self):\n",
    "        return self.seq_type\n",
    "    def info(self):\n",
    "        print (f\"SEQ: {self.seq}\" +\" \"+ f\"TYPE: {self.seq_type}\")\n",
    "        \n",
    "    '''Get the aphabet of the sequence type'''\n",
    "    def abc(self):\n",
    "        if(self.seq_type==\"DNA\"): \n",
    "          return \"ACGT\"\n",
    "        elif(self.seq_type==\"mRNA\"):\n",
    "          return \"ACGU\"\n",
    "        elif (self.seq_type==\"PROTEIN\"): \n",
    "          return \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "        else: \n",
    "          return None\n",
    "        \n",
    "    '''Check if sequence alphabet match dictionary'''\n",
    "    # calls abc function \n",
    "    def validate(self,verbose=False):\n",
    "        alp = self.abc()\n",
    "        res = True; i = 0\n",
    "        while (res and i < len(self.seq)):\n",
    "            if self.seq[i] not in alp: \n",
    "                res = False\n",
    "            else: i += 1\n",
    "        if(res):\n",
    "            if(verbose):\n",
    "                print(f'{self.seq_type} is valid')\n",
    "            return res\n",
    "        else:\n",
    "            if(verbose):\n",
    "                print(f'{self.seq_type} is invalid')\n",
    "            return res\n",
    "        \n",
    "    '''Transcription DNA -> RNA'''\n",
    "    def transcription(self):\n",
    "        if (self.seq_type == \"DNA\"):\n",
    "            return SQ(self.seq.replace(\"T\",\"U\"), \"mRNA\")\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    '''Reverse Compliment of DNA -> DNA'''\n",
    "    def reverse_comp(self):\n",
    "        \n",
    "        if (self.seq_type != \"DNA\"): \n",
    "            print('input not DNA')\n",
    "            return None\n",
    "    \n",
    "        lst_seq = ['A','T','G','C']\n",
    "        lst_comp = ['T','A','C','G']\n",
    "            \n",
    "        comp = ''\n",
    "        for char in self.seq:\n",
    "            ii=-1\n",
    "            for c in lst_seq:\n",
    "                ii+=1\n",
    "                if(char == c ):\n",
    "                    comp = lst_comp[ii] + comp\n",
    "            \n",
    "        return SQ(comp, \"DNA\")\n",
    "        \n",
    "    ''' Translates a DNA to aminoacid sequence '''\n",
    "    # using defined dictionary mapping\n",
    "    @staticmethod\n",
    "    def translate(seq,p0=0):\n",
    "        seq_aa = \"\"\n",
    "        for pos in range(p0,len(seq)-2,3):\n",
    "            cod = seq[pos:pos+3]\n",
    "            seq_aa += dic_map(map_id='codon',tid=cod)\n",
    "        return seq_aa\n",
    "    \n",
    "    '''Get All Possible open reading frames (ORF)'''\n",
    "    # store all possible collections of amino acid groups \n",
    "    # in all 6 frames\n",
    "    def frames(self):\n",
    "        res = []\n",
    "        for i in range(0,3):\n",
    "            res.append(self.translate(self.seq,i))\n",
    "        rc = self.reverse_comp()\n",
    "        for i in range(0,3):\n",
    "            res.append(self.translate(rc,i)) \n",
    "        return res\n",
    "    \n",
    "    '''Computes all possible proteins in an amino acid sequence in reading frame '''\n",
    "    # using the knowledge that it starts with M and ends with _, \n",
    "    # filter out rule breaking ORFs\n",
    "    @staticmethod\n",
    "    def all_proteins_RF(aa_seq):\n",
    "        # aa_seq -> converted ORF\n",
    "        current_prot = []\n",
    "        proteins = []\n",
    "        for aa in aa_seq:\n",
    "            if(aa == \"_\"):\n",
    "                if current_prot:\n",
    "                    for p in current_prot:\n",
    "                        proteins.append(p)\n",
    "                    current_prot = []\n",
    "            else:\n",
    "                if(aa == \"M\"):\n",
    "                    current_prot.append(\"\")\n",
    "                for i in range(len(current_prot)):\n",
    "                    current_prot[i] += aa\n",
    "        return proteins\n",
    "    \n",
    "    '''Computes all possible proteins for all ORF'''\n",
    "    # and sort them based on size\n",
    "    def ORF_protein(self, mins = 0):\n",
    "        \n",
    "        # order \n",
    "        def insert_prot_ord (prot, list_prots):\n",
    "            i = 0\n",
    "            while i < len(list_prots) and len(prot) < len(list_prots[i]):        \n",
    "                i += 1\n",
    "            list_prots.insert(i, prot)\n",
    "        \n",
    "        rfs = self.frames()  # get all ORF conversions\n",
    "        res = []\n",
    "        for rf in rfs:\n",
    "            prots = self.all_proteins_RF(rf) # return only protein cases\n",
    "            # additionally sort based on protein size\n",
    "            for p in prots: \n",
    "                if len(p) > mins: \n",
    "                    insert_prot_ord(p, res)\n",
    "        return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "contemporary-ghana",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:26.182491Z",
     "iopub.status.busy": "2021-09-08T13:53:26.181747Z",
     "iopub.status.idle": "2021-09-08T13:53:26.184878Z",
     "shell.execute_reply": "2021-09-08T13:53:26.185585Z"
    },
    "papermill": {
     "duration": 0.036014,
     "end_time": "2021-09-08T13:53:26.185830",
     "exception": false,
     "start_time": "2021-09-08T13:53:26.149816",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence: ATGACGGATCAGCCGCAAGCGGAATTGGCGTTTACGTACGATGCGCCGTAA | Sequence Type: DNA\n",
      "Sequence: TAATATGTTTTCGTTCATGCAGAGAGATTAAGGGTGTCTAATGAAGAAAAGTTCTATTGTGGCAACCATTATAACTATT | Sequence Type: DNA\n",
      "Sequence: MMMELQHQRLMALAGQLQLESLISAAPALSQQAVDQEWSYMDFLEHLLHEEKLARHQRKQAMYTRMAAFPAVKTFEEYDFTF | Sequence Type: PROTEIN\n",
      "Sequence: ATGAPQKQLQSLRSLSFIERNENIVLLGPSGVGKTHLAIAMGYEAVRAGIKVRFTTAADLLLQLSTAQRQGRYKTTLQRGVMAPRLLI | Sequence Type: PROTEIN\n"
     ]
    }
   ],
   "source": [
    "# Define sequences in string format\n",
    "seq1 = 'ATGACGGATCAGCCGCAAGCGGAATTGGCGTTTACGTACGATGCGCCGTAA'\n",
    "seq2 = 'TAATATGTTTTCGTTCATGCAGAGAGATTAAGGGTGTCTAATGAAGAAAAGTTCTATTGTGGCAACCATTATAACTATT'\n",
    "seqaa1 = 'MMMELQHQRLMALAGQLQLESLISAAPALSQQAVDQEWSYMDFLEHLLHEEKLARHQRKQAMYTRMAAFPAVKTFEEYDFTF'\n",
    "seqaa2 = 'ATGAPQKQLQSLRSLSFIERNENIVLLGPSGVGKTHLAIAMGYEAVRAGIKVRFTTAADLLLQLSTAQRQGRYKTTLQRGVMAPRLLI'\n",
    "\n",
    "# Define Two DNA sequences (instances)\n",
    "sq1 = SQ(seq1,'DNA')\n",
    "sq2 = SQ(seq2,'DNA')\n",
    "# define new protein sequence \n",
    "sq3 = SQ(seqaa1,'PROTEIN')\n",
    "sq4 = SQ(seqaa2,'PROTEIN')\n",
    "\n",
    "# Show class variables\n",
    "print(f'Sequence: {sq1.seq} | Sequence Type: {sq1.seq_type}')\n",
    "print(f'Sequence: {sq2.seq} | Sequence Type: {sq2.seq_type}')\n",
    "print(f'Sequence: {sq3.seq} | Sequence Type: {sq3.seq_type}')\n",
    "print(f'Sequence: {sq4.seq} | Sequence Type: {sq4.seq_type}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "chicken-district",
   "metadata": {
    "papermill": {
     "duration": 0.025475,
     "end_time": "2021-09-08T13:53:26.238628",
     "exception": false,
     "start_time": "2021-09-08T13:53:26.213153",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>3 |</b> SEQUENCE CLASS USAGE EXAMPLES\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "### **<span style='color:#FF5733'>3.1</span> | NUCLEOTIDE/AMINO ACID FREQUENCY**\n",
    "- Having defined all the <b>IUPAC Nucleotide</b> & <b>Amino acid</b> (From [TABLE](https://www.bioinformatics.org/sms/iupac.html)) in the dictionary <code>dic_map</code>, we may want to obtain the count for each each nucleotide or amino acid in the given sequence. \n",
    "- We can use the <code>.freq()</code> function to show the count/percentage of each of the alphabet or compare two sets of sequences with the addition of <b>compare</b>, eg. <code>freq(compare=None,show_id='perc')</code>, and changing between the <b>percentage (perc)</b> & <b>number count (count)</b> using <b>show_id</b>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "spoken-ensemble",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:26.293891Z",
     "iopub.status.busy": "2021-09-08T13:53:26.293228Z",
     "iopub.status.idle": "2021-09-08T13:53:27.349414Z",
     "shell.execute_reply": "2021-09-08T13:53:27.348902Z"
    },
    "papermill": {
     "duration": 1.085056,
     "end_time": "2021-09-08T13:53:27.349558",
     "exception": false,
     "start_time": "2021-09-08T13:53:26.264502",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>                            <div id=\"35cb8ae0-e7cf-4fb8-8064-87abb904d63a\" class=\"plotly-graph-div\" style=\"height:250px; width:100%;\"></div>            <script type=\"text/javascript\">                require([\"plotly\"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"35cb8ae0-e7cf-4fb8-8064-87abb904d63a\")) {                    Plotly.newPlot(                        \"35cb8ae0-e7cf-4fb8-8064-87abb904d63a\",                        [{\"marker\": {\"color\": \"rgb(26, 118, 255)\"}, \"name\": \"SEQ1\", \"orientation\": \"h\", \"text\": [\"0.255\", \"0.196\", \"0.314\", \"0.235\"], \"textposition\": \"outside\", \"type\": \"bar\", \"x\": [0.255, 0.196, 0.314, 0.235], \"y\": [\"Adenine\", \"Thymine\", \"Guanine\", \"Cytosine\"]}, {\"marker\": {\"color\": \"rgb(55, 83, 109)\"}, \"name\": \"SEQ2\", \"orientation\": \"h\", \"text\": [\"0.354\", \"0.329\", \"0.203\", \"0.114\"], \"textposition\": \"outside\", \"type\": \"bar\", \"x\": [0.354, 0.329, 0.203, 0.114], \"y\": [\"Thymine\", \"Adenine\", \"Guanine\", \"Cytosine\"]}],                        {\"font\": {\"family\": \"sans-serif\", \"size\": 12}, \"height\": 250, \"margin\": {\"b\": 10, \"l\": 40, \"r\": 40, \"t\": 50}, \"template\": {\"data\": {\"bar\": [{\"error_x\": {\"color\": \"#2a3f5f\"}, \"error_y\": {\"color\": \"#2a3f5f\"}, \"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"bar\"}], \"barpolar\": [{\"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"barpolar\"}], \"carpet\": [{\"aaxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"baxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"type\": \"carpet\"}], \"choropleth\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"choropleth\"}], \"contour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"contour\"}], \"contourcarpet\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"contourcarpet\"}], \"heatmap\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmap\"}], \"heatmapgl\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmapgl\"}], \"histogram\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"histogram\"}], \"histogram2d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2d\"}], \"histogram2dcontour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2dcontour\"}], \"mesh3d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"mesh3d\"}], \"parcoords\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"parcoords\"}], \"pie\": [{\"automargin\": true, \"type\": \"pie\"}], \"scatter\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter\"}], \"scatter3d\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter3d\"}], \"scattercarpet\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattercarpet\"}], \"scattergeo\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergeo\"}], \"scattergl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergl\"}], \"scattermapbox\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattermapbox\"}], \"scatterpolar\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolar\"}], \"scatterpolargl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolargl\"}], \"scatterternary\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterternary\"}], \"surface\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"surface\"}], \"table\": [{\"cells\": {\"fill\": {\"color\": \"#EBF0F8\"}, \"line\": {\"color\": \"white\"}}, \"header\": {\"fill\": {\"color\": \"#C8D4E3\"}, \"line\": {\"color\": \"white\"}}, \"type\": \"table\"}]}, \"layout\": {\"annotationdefaults\": {\"arrowcolor\": \"#2a3f5f\", \"arrowhead\": 0, \"arrowwidth\": 1}, \"autotypenumbers\": \"strict\", \"coloraxis\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"colorscale\": {\"diverging\": [[0, \"#8e0152\"], [0.1, \"#c51b7d\"], [0.2, \"#de77ae\"], [0.3, \"#f1b6da\"], [0.4, \"#fde0ef\"], [0.5, \"#f7f7f7\"], [0.6, \"#e6f5d0\"], [0.7, \"#b8e186\"], [0.8, \"#7fbc41\"], [0.9, \"#4d9221\"], [1, \"#276419\"]], \"sequential\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"sequentialminus\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]]}, \"colorway\": [\"#636efa\", \"#EF553B\", \"#00cc96\", \"#ab63fa\", \"#FFA15A\", \"#19d3f3\", \"#FF6692\", \"#B6E880\", \"#FF97FF\", \"#FECB52\"], \"font\": {\"color\": \"#2a3f5f\"}, \"geo\": {\"bgcolor\": \"white\", \"lakecolor\": \"white\", \"landcolor\": \"white\", \"showlakes\": true, \"showland\": true, \"subunitcolor\": \"#C8D4E3\"}, \"hoverlabel\": {\"align\": \"left\"}, \"hovermode\": \"closest\", \"mapbox\": {\"style\": \"light\"}, \"paper_bgcolor\": \"white\", \"plot_bgcolor\": \"white\", \"polar\": {\"angularaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"radialaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}}, \"scene\": {\"xaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"yaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"zaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}}, \"shapedefaults\": {\"line\": {\"color\": \"#2a3f5f\"}}, \"ternary\": {\"aaxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"baxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"caxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}}, \"title\": {\"x\": 0.05}, \"xaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}, \"yaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}}}, \"title\": {\"text\": \"<b>DNA SEQUENCE CONTENT</b>\"}},                        {\"responsive\": true}                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('35cb8ae0-e7cf-4fb8-8064-87abb904d63a');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })                };                });            </script>        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# compare content percentage (default)\n",
    "sq1.freq(sq2,fheight=250)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "following-scheme",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:27.436616Z",
     "iopub.status.busy": "2021-09-08T13:53:27.425927Z",
     "iopub.status.idle": "2021-09-08T13:53:27.446857Z",
     "shell.execute_reply": "2021-09-08T13:53:27.446208Z"
    },
    "papermill": {
     "duration": 0.070047,
     "end_time": "2021-09-08T13:53:27.447002",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.376955",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>                            <div id=\"9b5cc245-5793-4407-ba55-558d6c874f94\" class=\"plotly-graph-div\" style=\"height:500px; width:100%;\"></div>            <script type=\"text/javascript\">                require([\"plotly\"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"9b5cc245-5793-4407-ba55-558d6c874f94\")) {                    Plotly.newPlot(                        \"9b5cc245-5793-4407-ba55-558d6c874f94\",                        [{\"marker\": {\"color\": \"rgb(26, 118, 255)\"}, \"name\": \"SEQ1\", \"orientation\": \"h\", \"text\": [\"7\", \"8\", \"11\", \"9\", \"4\", \"4\", \"11\", \"1\", \"4\", \"1\", \"2\", \"2\", \"3\", \"1\", \"3\", \"5\", \"3\", \"3\", \"0\", \"0\"], \"textposition\": \"outside\", \"type\": \"bar\", \"x\": [7, 8, 11, 9, 4, 4, 11, 1, 4, 1, 2, 2, 3, 1, 3, 5, 3, 3, 0, 0], \"y\": [\"Methionine\", \"Glutamic Acid\", \"Lysine\", \"Glutamine\", \"Histidine\", \"Arginine\", \"Alanine\", \"Glycine\", \"Serine\", \"Isoleucine\", \"Proline\", \"Valine\", \"Aspartic Acid\", \"Tryptophan\", \"Tryosine\", \"Phenylalanine\", null, \"Threonine\", \"Cysteine\", \"Asparagine\"]}],                        {\"font\": {\"family\": \"sans-serif\", \"size\": 12}, \"height\": 500, \"margin\": {\"b\": 10, \"l\": 40, \"r\": 40, \"t\": 50}, \"template\": {\"data\": {\"bar\": [{\"error_x\": {\"color\": \"#2a3f5f\"}, \"error_y\": {\"color\": \"#2a3f5f\"}, \"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"bar\"}], \"barpolar\": [{\"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"barpolar\"}], \"carpet\": [{\"aaxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"baxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"type\": \"carpet\"}], \"choropleth\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"choropleth\"}], \"contour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"contour\"}], \"contourcarpet\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"contourcarpet\"}], \"heatmap\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmap\"}], \"heatmapgl\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmapgl\"}], \"histogram\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"histogram\"}], \"histogram2d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2d\"}], \"histogram2dcontour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2dcontour\"}], \"mesh3d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"mesh3d\"}], \"parcoords\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"parcoords\"}], \"pie\": [{\"automargin\": true, \"type\": \"pie\"}], \"scatter\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter\"}], \"scatter3d\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter3d\"}], \"scattercarpet\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattercarpet\"}], \"scattergeo\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergeo\"}], \"scattergl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergl\"}], \"scattermapbox\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattermapbox\"}], \"scatterpolar\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolar\"}], \"scatterpolargl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolargl\"}], \"scatterternary\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterternary\"}], \"surface\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"surface\"}], \"table\": [{\"cells\": {\"fill\": {\"color\": \"#EBF0F8\"}, \"line\": {\"color\": \"white\"}}, \"header\": {\"fill\": {\"color\": \"#C8D4E3\"}, \"line\": {\"color\": \"white\"}}, \"type\": \"table\"}]}, \"layout\": {\"annotationdefaults\": {\"arrowcolor\": \"#2a3f5f\", \"arrowhead\": 0, \"arrowwidth\": 1}, \"autotypenumbers\": \"strict\", \"coloraxis\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"colorscale\": {\"diverging\": [[0, \"#8e0152\"], [0.1, \"#c51b7d\"], [0.2, \"#de77ae\"], [0.3, \"#f1b6da\"], [0.4, \"#fde0ef\"], [0.5, \"#f7f7f7\"], [0.6, \"#e6f5d0\"], [0.7, \"#b8e186\"], [0.8, \"#7fbc41\"], [0.9, \"#4d9221\"], [1, \"#276419\"]], \"sequential\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"sequentialminus\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]]}, \"colorway\": [\"#636efa\", \"#EF553B\", \"#00cc96\", \"#ab63fa\", \"#FFA15A\", \"#19d3f3\", \"#FF6692\", \"#B6E880\", \"#FF97FF\", \"#FECB52\"], \"font\": {\"color\": \"#2a3f5f\"}, \"geo\": {\"bgcolor\": \"white\", \"lakecolor\": \"white\", \"landcolor\": \"white\", \"showlakes\": true, \"showland\": true, \"subunitcolor\": \"#C8D4E3\"}, \"hoverlabel\": {\"align\": \"left\"}, \"hovermode\": \"closest\", \"mapbox\": {\"style\": \"light\"}, \"paper_bgcolor\": \"white\", \"plot_bgcolor\": \"white\", \"polar\": {\"angularaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"radialaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}}, \"scene\": {\"xaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"yaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"zaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}}, \"shapedefaults\": {\"line\": {\"color\": \"#2a3f5f\"}}, \"ternary\": {\"aaxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"baxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"caxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}}, \"title\": {\"x\": 0.05}, \"xaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}, \"yaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}}}, \"title\": {\"text\": \"<b>PROTEIN SEQUENCE CONTENT</b>\"}},                        {\"responsive\": true}                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('9b5cc245-5793-4407-ba55-558d6c874f94');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })                };                });            </script>        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# show the amino acid content of the protein\n",
    "sq3.freq(show_id='count',fheight=500)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "consistent-breed",
   "metadata": {
    "papermill": {
     "duration": 0.028044,
     "end_time": "2021-09-08T13:53:27.503007",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.474963",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>3.2</span> | GC CONTENT**\n",
    "From GC-CONTENT | [WIKIPEDIA](https://en.wikipedia.org/wiki/GC-content)\n",
    "\n",
    "> In molecular biology and genetics, GC-content (or guanine-cytosine content) is the percentage of nitrogenous bases in a DNA or RNA molecule that are either <b>guanine</b> (G) or <b>cytosine</b> (C). This measure indicates the proportion of G and C bases out of an implied four total bases, also including adenine and thymine in DNA and adenine and uracil in RNA.\n",
    "\n",
    "From GC-CONTENT | [SCIENCEDIRECT](https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/gc-content)\n",
    "> GC content is strongly correlated with biological features of genome organization such as distribution of various classes of repeated elements, gene density, level and tissue-specificity of transcription, and mutation rate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "certified-hybrid",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:27.564358Z",
     "iopub.status.busy": "2021-09-08T13:53:27.563658Z",
     "iopub.status.idle": "2021-09-08T13:53:27.566822Z",
     "shell.execute_reply": "2021-09-08T13:53:27.566191Z"
    },
    "papermill": {
     "duration": 0.036012,
     "end_time": "2021-09-08T13:53:27.566970",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.530958",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC-Content Sequence 1: 54.9%\n",
      "GC-Content Sequence 2: 31.65%\n"
     ]
    }
   ],
   "source": [
    "print(f'GC-Content Sequence 1: {round(sq1.gc()*100,3)}%')\n",
    "print(f'GC-Content Sequence 2: {sq2.gc()*100}%')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hazardous-study",
   "metadata": {
    "papermill": {
     "duration": 0.027917,
     "end_time": "2021-09-08T13:53:27.623327",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.595410",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>3.3</span> | SEQUENCE VALIDITY**\n",
    "- A sequence is valid if its alphabet corresponds to a specific set of code, class <b>SQ</b> uses the [IUPAC](https://www.bioinformatics.org/sms2/iupac.html) code standard.\n",
    "- We can check whether a sequence is valid or not by using the <code>.validate()</code> function, returning a logical output (True/False)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "hourly-office",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:27.683385Z",
     "iopub.status.busy": "2021-09-08T13:53:27.682660Z",
     "iopub.status.idle": "2021-09-08T13:53:27.688676Z",
     "shell.execute_reply": "2021-09-08T13:53:27.689417Z"
    },
    "papermill": {
     "duration": 0.037988,
     "end_time": "2021-09-08T13:53:27.689667",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.651679",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: MMMELQHQRLMALAGQLQLESLISAAPALSQQAVDQEWSYMDFLEHLLHEEKLARHQRKQAMYTRMAAFPAVKTFEEYDFTF is valid: True\n"
     ]
    }
   ],
   "source": [
    "print(f'sequence: {sq3.seq} is valid: {sq3.validate()}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compliant-personal",
   "metadata": {
    "papermill": {
     "duration": 0.02933,
     "end_time": "2021-09-08T13:53:27.749491",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.720161",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>3.4</span> | TRANSCRIPTION**\n",
    "- The mRNA is created as the two complementary strands are split, and is complement to one of the strands.\n",
    "- Transcription of <b>DNA sequence</b> can be obtained using the <code>.transcription()</code> function, where the <b>T</b> is replaced by <b>U</b> in the sequence string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "authorized-activation",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:27.814359Z",
     "iopub.status.busy": "2021-09-08T13:53:27.813525Z",
     "iopub.status.idle": "2021-09-08T13:53:27.816933Z",
     "shell.execute_reply": "2021-09-08T13:53:27.817438Z"
    },
    "papermill": {
     "duration": 0.03927,
     "end_time": "2021-09-08T13:53:27.817612",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.778342",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA sequence: UAAUAUGUUUUCGUUCAUGCAGAGAGAUUAAGGGUGUCUAAUGAAGAAAAGUUCUAUUGUGGCAACCAUUAUAACUAUU\n"
     ]
    }
   ],
   "source": [
    "mrna_sq2 = sq2.transcription()\n",
    "print(f'mRNA sequence: {mrna_sq2.seq}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "piano-mistress",
   "metadata": {
    "papermill": {
     "duration": 0.028764,
     "end_time": "2021-09-08T13:53:27.875652",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.846888",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>3.5</span> | REVERSE COMPLEMENT**\n",
    "- DNA has <b>two complementary strands</b>. \n",
    "- Due to the complementarity of the DNA strands, usually only one of the strands is provided in a sequence file obtained from databases.\n",
    "- The second strand to the input <b>DNA sequence</b> can be obtained by calling the <code>.reverse_comp()</code> function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "subsequent-affair",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:27.940951Z",
     "iopub.status.busy": "2021-09-08T13:53:27.940069Z",
     "iopub.status.idle": "2021-09-08T13:53:27.943970Z",
     "shell.execute_reply": "2021-09-08T13:53:27.944427Z"
    },
    "papermill": {
     "duration": 0.039321,
     "end_time": "2021-09-08T13:53:27.944618",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.905297",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEQ: TTACGGCGCATCGTACGTAAACGCCAATTCCGCTTGCGGCTGATCCGTCAT TYPE: DNA\n"
     ]
    }
   ],
   "source": [
    "rev_comp_sq1 = sq1.reverse_comp()\n",
    "rev_comp_sq1.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "instructional-province",
   "metadata": {
    "papermill": {
     "duration": 0.02963,
     "end_time": "2021-09-08T13:53:28.004029",
     "exception": false,
     "start_time": "2021-09-08T13:53:27.974399",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>3.6</span> | TRANSLATION**\n",
    "#### <b><span style='color:#FF5733'>OVERVIEW</span></b>\n",
    "- The background to what happens during the translation process was already outlined in <code>1.3</code>, however it makes sense to expand on theory a little more.\n",
    "- Proteins are synthesised by creating <b>chains of aminoacids</b>, according to information contained in the <b>messenger RNA (mRNA)</b> in a process called <b>translation</b>.\n",
    "\n",
    "#### <b><span style='color:#FF5733'>START & STOP CODONS</span></b>\n",
    "- <b>Translation</b> of a protein always begins with a specific codon <b>ATG</b> -> <b>M (Methionine)</b>, which is always the first amino in the protein.\n",
    "- <b>Translation</b> process terminates when a stop codon is found; <b>TAA</b>,<b>TAG</b>,<b>TGA</b> -> <code>_</code>.\n",
    "- An example sequence where we know exactly where the <b>start</b> and <b>termination codons</b> are:\n",
    "    - <b>ATG</b><code>ACGGATCAGCCGCAAGCGGAATTGGCGTTTACGTACGATGCGCCG</code><b>TAA</b>\n",
    "    - We can note it starts with <b>ATG</b> & ends with one of the three stop codons <b>TAA</b>.\n",
    "- As the sequence follows the rule of start and ending codon, in such a case we can use the <b>staticmethod</b> defined in SQ; <code>.translate(p0=0)</code> directly.\n",
    "\n",
    "#### <b><span style='color:#FF5733'>OPEN READING FRAME</span></b>\n",
    "- A <b>reading frame</b> is a way of dividing the DNA sequence into a set of consecutive, <b>non-overlapping triplet nucleotides</b> (possible codons) (using dictionary mapping).\n",
    "- A given sequence has 3 possible reading frames, first, second and third nucleotide positions. In addition, considering there is another complementary strand, we should compute the only 3 frames corresponding to the reverse compliment.\n",
    "- In many cases, given a DNA sequence, <b>we don't know in advance where the coding regions are</b>, especially when dealing with complete sequences.\n",
    "- In such cases, we need to <b>scan the DNA sequence for the coding region</b>. First, we need to divide and compute these reading frames (6 in total). The <code>frames</code> function stores the converted 6 converted amino acid strings in a list.\n",
    "\n",
    "<b>OPEN READING FRAME</b> | [genomove.gov](https://www.genome.gov/genetics-glossary/Open-Reading-Frame)\n",
    "\n",
    "> An open reading frame is a portion of a DNA molecule that, when translated into amino acids, contains no stop codons. The genetic code reads DNA sequences in groups of three base pairs, which means that a double-stranded DNA molecule can read in any of six possible reading frames--three in the forward direction and three in the reverse. A long open reading frame is likely part of a gene.\n",
    "\n",
    "<center>\n",
    "\n",
    "| Open Reading Frames (ORF)[[1]](https://www.genome.gov/genetics-glossary/Open-Reading-Frame) |\n",
    "| - |\n",
    "|<img src=\"https://www.genome.gov/sites/default/files/tg/en/illustration/open_reading_frame.jpg\" alt=\"Drawing\" style=\"width:700px;\"/> |\n",
    "    \n",
    "</center>\n",
    "\n",
    "#### <b><span style='color:#FF5733'>IDEAL CASE (P=0)</span></b>\n",
    "- Let's try one of the reading frames at the start of the sequence (<b>p0=0</b>), which we know follows the correct rules observed in life.\n",
    "- Let's also consider all other possible ORFS & get the final protein found in the DNA sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "neither-bones",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.071664Z",
     "iopub.status.busy": "2021-09-08T13:53:28.070926Z",
     "iopub.status.idle": "2021-09-08T13:53:28.075499Z",
     "shell.execute_reply": "2021-09-08T13:53:28.074964Z"
    },
    "papermill": {
     "duration": 0.040069,
     "end_time": "2021-09-08T13:53:28.075637",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.035568",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correct ORF: MTDQPQAELAFTYDAP_\n",
      "All ORF:\n",
      "MTDQPQAELAFTYDAP_    DGSAASGIGVYVRCAV     YGASYVNANSACG_SV     \n",
      "_RISRKRNWRLRTMRR     LRRIVRKRQFRLRLIRH    TAHRT_TPIPLAADPS     \n"
     ]
    }
   ],
   "source": [
    "# Sequence with initial codon coinciding with start codon &\n",
    "# end w/ end codon (ATG & TAA respectively)\n",
    "seq = 'ATGACGGATCAGCCGCAAGCGGAATTGGCGTTTACGTACGATGCGCCGTAA'\n",
    "strand = SQ(seq=seq,seq_type='DNA')\n",
    "print(f'Correct ORF: {strand.translate(strand.seq,p0=0)}')\n",
    "print(f'All ORF:')\n",
    "lst_ORF = strand.frames()\n",
    "lstcol(lst_ORF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "nuclear-option",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.144640Z",
     "iopub.status.busy": "2021-09-08T13:53:28.143978Z",
     "iopub.status.idle": "2021-09-08T13:53:28.147221Z",
     "shell.execute_reply": "2021-09-08T13:53:28.147672Z"
    },
    "papermill": {
     "duration": 0.041997,
     "end_time": "2021-09-08T13:53:28.147868",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.105871",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MTDQPQAELAFTYDAP']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only one of the ORFs meets the requirement, so only one \n",
    "# protein is found \n",
    "strand.ORF_protein()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "potential-caribbean",
   "metadata": {
    "papermill": {
     "duration": 0.030722,
     "end_time": "2021-09-08T13:53:28.209258",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.178536",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>4 |</b> LOADING FILES CONTAINING SEQUENCES\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "### <b><span style='color:#FF5733'>4.1</span> | REAL SEQUENCES</b>\n",
    "- It's quite straightforward to work with very short sequences; simply using the python <code>input()</code> function.\n",
    "- Most realistic application of bioinformatics certainly involve sequences that are <b>too big to type out</b> & thus have to be read from files.\n",
    "- On top of that, databases already store specific formats which contain information about a sequence and the sequence itself; one of such formats is <b>FASTA</b>.\n",
    "\n",
    "### <b><span style='color:#FF5733'>4.2</span> | THE FASTA FORMAT</b>\n",
    "\n",
    "Snipplet from [FASTA format](https://en.wikipedia.org/wiki/FASTA_format):\n",
    ">In bioinformatics and biochemistry, the FASTA format is a text-based format for representing either <b>nucleotide sequences</b> or <b>amino acid (protein) sequences</b>, in which nucleotides or amino acids are represented using single-letter codes. The format also allows for sequence names and comments to precede the sequences. The format originates from the FASTA software package, but has now become a near universal standard in the field of bioinformatics.\n",
    "\n",
    "#### <b><span style='color:#FF5733'>OVERVIEW</span></b>\n",
    "- The format is commonly used to store nucleotide or protein sequences, it's less rigorous, usually containing only the <b>sequence</b>, and <b>name/header</b> only.\n",
    "- The file extension typically changed based on the <b>sequence type</b> content of the file, defined below in the dictionary <code>FASTA_dic</code>.\n",
    "- The format can contain any number of sequences, each starting with the symbol: <b>></b>.\n",
    "- The <b>name/header</b>, defined after the symbol <b>></b> contains an origin identifier (<b>NCBI identifiers</b>), defined in the the dictionary <code>identifiers_dic</code>.\n",
    "\n",
    "> The NCBI defined a standard for the unique identifier used for the sequence (SeqID) in the header line. This allows a sequence that was obtained from a database to be labelled with a reference to its database record. The database identifier format is understood by the NCBI tools like makeblastdb and table2asn. The following list describes the NCBI FASTA defined format for sequence identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "entitled-world",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.273726Z",
     "iopub.status.busy": "2021-09-08T13:53:28.273042Z",
     "iopub.status.idle": "2021-09-08T13:53:28.279199Z",
     "shell.execute_reply": "2021-09-08T13:53:28.279738Z"
    },
    "papermill": {
     "duration": 0.039951,
     "end_time": "2021-09-08T13:53:28.279935",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.239984",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# NCBI identifiers\n",
    "identifiers_dic = {'lcl':'local(nodb)','bbs':'GenInfo backbone seqid',\n",
    "                   'bbm':'GenInfo backbone moltype','gim':'GenInfo import ID',\n",
    "                   'gb':'GenBank','emb':'EMBL','pir':'PIR','sp':'SWISS-PROT',\n",
    "                   'pat':'patent','pgp':'pre-grant patent','ref':'RefSeq',\n",
    "                   'gnl':'general database reference','prf':'PRF','pdb':'PDB',\n",
    "                   'gi':'GenInfo integrated database','dbj':'DDBJ'}\n",
    "\n",
    "# FASTA formats\n",
    "FASTA_dic = {'fa':'generic','fasta':'generic','fna':'nucleic acid',\n",
    "             'ffn':'nucleotide of gene regions','faa':'amino acid',\n",
    "             'frn':'non-coding RNA'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tight-milton",
   "metadata": {
    "papermill": {
     "duration": 0.030405,
     "end_time": "2021-09-08T13:53:28.340970",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.310565",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### **<span style='color:#FF5733'>4.3</span> | READ SEQUENCE CLASS, read_seq()**\n",
    "- Class <b>read_seq</b> defines function a <code>read_FASTA</code>, that can be used to read <b>FASTA</b> based formats and is integrated into the previously defined general purpose sequence class.\n",
    "- More sophisticated formats can also be incorporated into the current class. [BioPython GenBank Dataset](https://www.kaggle.com/shtrausslearning/biopython-genbank) contains other formats as well.\n",
    "- The class automatically should detect the class type and call the corresponding class, storing the sequence upon instantiation. \n",
    "- Since the constructor can't return anything upon instantiation, <b>get_sq</b> is used to return a SQ class instance for each of the read sequences in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "authentic-costa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.416662Z",
     "iopub.status.busy": "2021-09-08T13:53:28.415999Z",
     "iopub.status.idle": "2021-09-08T13:53:28.418608Z",
     "shell.execute_reply": "2021-09-08T13:53:28.419288Z"
    },
    "papermill": {
     "duration": 0.047899,
     "end_time": "2021-09-08T13:53:28.419465",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.371566",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Class to read different files and store info only\n",
    "class read_seq(SQ):\n",
    "    \n",
    "    def __init__(self,name):\n",
    "        self.name = name\n",
    "        self.format = name.split('.')[1]\n",
    "        if(self.format in FASTA_dic):      # if one of the fasta formats\n",
    "            self.read_FASTA(self.name)\n",
    "\n",
    "    # read FASTA format\n",
    "    def read_FASTA(self,filename):\n",
    "\n",
    "        tseq = None; self.lst_seq = []     # list of sequences\n",
    "        thead = None; self.lst_header = [] # list of sequence identifications\n",
    "        ff = FASTA_dic[filename.split('.')[1]]\n",
    "        file = open(filename,'r')\n",
    "\n",
    "        for line in file:\n",
    "            if(search(\">.*\", line)): # get lines w/ >\n",
    "                    if(tseq != None and thead != None and tseq != \"\"):\n",
    "                        self.lst_seq.append(tseq)\n",
    "                    thead = line; self.lst_header.append(line)              \n",
    "                    tseq = \"\"\n",
    "            else:\n",
    "                if(tseq == None):\n",
    "                    return None\n",
    "                else: \n",
    "                    tseq += sub(\"\\s\",\"\",line)\n",
    "\n",
    "        if(tseq != None and thead != None and tseq != \"\"):\n",
    "            self.lst_seq.append(tseq)\n",
    "            \n",
    "        print(f'READ -> FASTA [{ff}] | #SEQ: {len(self.lst_seq)}')\n",
    "        file.close()\n",
    "        \n",
    "    # get read sequences\n",
    "    def get_sq(self):\n",
    "        lst_out = []\n",
    "        if(len(self.lst_seq) > 1):\n",
    "            for i in range(0,len(self.lst_seq)):\n",
    "                lst_types = ['DNA','mRNA','PROTEIN']\n",
    "                for check in lst_types:\n",
    "                    if(SQ(self.lst_seq[i],check).validate()):\n",
    "                        lst_out.append(SQ(self.lst_seq[i],check)) \n",
    "            return lst_out\n",
    "        else:\n",
    "            lst_types = ['DNA','mRNA','PROTEIN']\n",
    "            for check in lst_types:\n",
    "                if(SQ(self.lst_seq[0],check).validate()): # if valid sq\n",
    "                    return SQ(self.lst_seq[0],check)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forced-copper",
   "metadata": {
    "papermill": {
     "duration": 0.030873,
     "end_time": "2021-09-08T13:53:28.481380",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.450507",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "- Fetched (Validate & Set) <b>set of sequences</b> are stored in the output variable in a <b>list</b> format; containing a set of <b>SQ</b> instances for each sequence in the file, whilst FASTA files containing only a <b>single sequence</b> are class instances.\n",
    "- Two examples are shown below, reading files: <code>file_faa</code> & <code>file_fna</code>, containing a <b>set of protein sequences</b> and a <b>single nucleotide sequence</b> resepectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "figured-requirement",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.548443Z",
     "iopub.status.busy": "2021-09-08T13:53:28.547642Z",
     "iopub.status.idle": "2021-09-08T13:53:28.570063Z",
     "shell.execute_reply": "2021-09-08T13:53:28.570779Z"
    },
    "papermill": {
     "duration": 0.058719,
     "end_time": "2021-09-08T13:53:28.571027",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.512308",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "READ -> FASTA [amino acid] | #SEQ: 10\n",
      "READ -> FASTA [nucleic acid] | #SEQ: 1\n"
     ]
    }
   ],
   "source": [
    "# define pathway to FASTA file\n",
    "file_faa = '/kaggle/input/biopython-genbank/NC_005816.faa'\n",
    "file_fna = '/kaggle/input/biopython-genbank/NC_005816.fna'\n",
    "\n",
    "# fetch sequence from file and store each in sequence class, SQ\n",
    "col_seq_aa = read_seq(file_faa).get_sq()\n",
    "col_seq_n = read_seq(file_fna).get_sq()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "appreciated-waste",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.639897Z",
     "iopub.status.busy": "2021-09-08T13:53:28.639234Z",
     "iopub.status.idle": "2021-09-08T13:53:28.642654Z",
     "shell.execute_reply": "2021-09-08T13:53:28.643273Z"
    },
    "papermill": {
     "duration": 0.03995,
     "end_time": "2021-09-08T13:53:28.643452",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.603502",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence: MMMELQHQRLMALAGQLQLESLISAAPALSQQAVDQEWSYMDFLEHLLHEEKLARHQRKQAMYTRMAAFPAVKTFEEYDFTFATGAPQKQLQSLRSLSFIERNENIVLLGPSGVGKTHLAIAMGYEAVRAGIKVRFTTAADLLLQLSTAQRQGRYKTTLQRGVMAPRLLIIDEIGYLPFSQEEAKLFFQVIAKRYEKSAMILTSNLPFGQWDQTFAGDAALTSAMLDRILHHSHVVQIKGESYRLRQKRKAGVIAEANPE\n",
      "sequence type: PROTEIN\n"
     ]
    }
   ],
   "source": [
    "print(f'sequence: {col_seq_aa[1].seq}')\n",
    "print(f'sequence type: {col_seq_aa[1].seq_type}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "awful-hospital",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.711298Z",
     "iopub.status.busy": "2021-09-08T13:53:28.710640Z",
     "iopub.status.idle": "2021-09-08T13:53:28.717972Z",
     "shell.execute_reply": "2021-09-08T13:53:28.717409Z"
    },
    "papermill": {
     "duration": 0.042015,
     "end_time": "2021-09-08T13:53:28.718124",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.676109",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Sequences stored: 10\n",
      "List of Sequences Type: <class 'list'>\n",
      "List Content Type: <class '__main__.SQ'>\n",
      "Single Fetched Sequence Type: <class '__main__.SQ'>\n"
     ]
    }
   ],
   "source": [
    "print(f'Number of Sequences stored: {len(col_seq_aa)}'); \n",
    "print(f'List of Sequences Type: {type(col_seq_aa)}')\n",
    "print(f'List Content Type: {type(col_seq_aa[0])}')\n",
    "print(f'Single Fetched Sequence Type: {type(col_seq_n)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "lucky-nylon",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.791919Z",
     "iopub.status.busy": "2021-09-08T13:53:28.791231Z",
     "iopub.status.idle": "2021-09-08T13:53:28.795467Z",
     "shell.execute_reply": "2021-09-08T13:53:28.794950Z"
    },
    "papermill": {
     "duration": 0.044739,
     "end_time": "2021-09-08T13:53:28.795605",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.750866",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proteins in Sequence col_seq_subset:\n",
      "MNALRMVIPPRYPWSLISRAITVAGIL    MLILTKSRQR                     \n",
      "MVIPPRYPWSLISRAITVAGIL         MP                             \n",
      "MNTGIIFVNASPMLILTKSRQR         MR                             \n",
      "MSIGDALTNIIPVFIQE                                             \n"
     ]
    }
   ],
   "source": [
    "# Select only a subset as the strand is a little too big (arbitrarily selected)\n",
    "col_seq_subset = SQ(col_seq_n.seq[100:500],col_seq_n.seq_type)\n",
    "# get all proteins with a length of more than 1\n",
    "proteins = col_seq_subset.ORF_protein(mins=1)\n",
    "\n",
    "# list all found proteins above a length of 2 \n",
    "print('Proteins in Sequence col_seq_subset:')\n",
    "lstcol(proteins,2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "retained-nowhere",
   "metadata": {
    "papermill": {
     "duration": 0.032261,
     "end_time": "2021-09-08T13:53:28.862385",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.830124",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>5 |</b> COVID-19: PROTEINS IDENTIFICATION\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "### <b><span style='color:#FF5733'>5.1</span> | REAL WORLD EXAMPLE : PROTEIN IDENTIFICATION</b>\n",
    "- Having defined a class that can <b>read</b>, <b>store</b> & <b>derive proteins</b> from a sequence file, we can use a database to obtain a real sequence, which should be quite familiar. \n",
    "- Let's use the Coronavirus sequence, already uploaded to [Kaggle](https://www.kaggle.com/paultimothymooney/coronavirus-genome-sequence) & [original source](https://www.ncbi.nlm.nih.gov/nuccore/NC_045512).\n",
    "\n",
    "#### <b><span style='color:#FF5733'>BLAST</span></b>\n",
    "\n",
    "- Our goal here will be to (1) get the proteins encoded in the genome & (2) learn something about them. \n",
    "- To do this, we'll be using a very useful resource that uses [BLAST (Basic Local Alignment Search Tool)](https://blast.ncbi.nlm.nih.gov/Blast.cgi). \n",
    "- We will look into <b>Local Sequence Alignment</b> in another notebook, however for the time being, a protein sequence is compared to others already found in a database & is compared with a corresponding <b>similarity score</b>.\n",
    "- To use the databse, we'll first need to (1) get the proteins & (2) simply search for matching results in the database (we'll need to copy & past our resultant protein sequences)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "spoken-parish",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:28.932592Z",
     "iopub.status.busy": "2021-09-08T13:53:28.931856Z",
     "iopub.status.idle": "2021-09-08T13:53:28.950855Z",
     "shell.execute_reply": "2021-09-08T13:53:28.950278Z"
    },
    "papermill": {
     "duration": 0.055875,
     "end_time": "2021-09-08T13:53:28.951000",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.895125",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "READ -> FASTA [nucleic acid] | #SEQ: 1\n",
      "Sequence Header: >MN908947.3 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      "\n",
      "Sequence Length: 96\n"
     ]
    }
   ],
   "source": [
    "# Read FASTA format containing Covid Genome\n",
    "virus_fna = '/kaggle/input/coronavirus-genome-sequence/MN908947.fna'\n",
    "virus_n = read_seq(virus_fna) # read and store FNA data \n",
    "print(f'Sequence Header: {virus_n.lst_header[0]}')\n",
    "print(f'Sequence Length: {len(virus_n.lst_header[0])}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eight-aurora",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:29.032968Z",
     "iopub.status.busy": "2021-09-08T13:53:29.032226Z",
     "iopub.status.idle": "2021-09-08T13:53:29.036381Z",
     "shell.execute_reply": "2021-09-08T13:53:29.036863Z"
    },
    "papermill": {
     "duration": 0.051894,
     "end_time": "2021-09-08T13:53:29.037030",
     "exception": false,
     "start_time": "2021-09-08T13:53:28.985136",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 29903 nucleotides\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAACGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGCAGGCTGCTTACGGTTTCGTCCGTGTTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCGAAAGGTAAGATGGAGAGCCTTGTCCCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTTTACAGGTTCGCGACGTGCTCGTACGTGGCTTTGGAGACTCCGTGGAGGAGGTCTTATCAGAGGCACGTCAACATCTTAAAGATGGCACTTGTGGCTTAGTAGAAGTTGAAAAAGGCGTTTTGCCTCAACTTGAACAGCCCTATGTGTTCATCAAACGTTCGGATGCTCGAACTGCACCTCATGGTCATGTTATGGTTGAGCTGGTAGCAGAACTCGAAGGCATTCAGTACGGTCGTAGTGGTGAGACACTTGGTGTCCTTGTCCCTCATGTGGGCGAAATACCAGTGGCTTACCGCAAGGTTCTTCTTCGTAAGAACGGTAATAAAGGAGCTGGTGGCCATAGTTACGGCGCCGATCTAAAGTCATTTGACTTAGGCGACGAGCTTGGCACTGATCCTTATGAAGATTTTCAAGAAAACTGGAACACTAAACATAGCAGTGGTGTTACCCGTGAACTCATGCGTGAGCTTAACGGAGGGGCATACACTCGCTATGTCGATAACAACTTCTGTGGCCCTGATGGCTACCCTCTTGAGTGCATTAAAGACCTTCTAGCACGTGCTGGTAAAGCTTCATGCACTTTGTCCGAACAACTGGACTTTATTGACACTAAGAGGGGTGTATACTGCTGCCGTGAACATGAGCATGAAATTGCTTGGTACACGGAACGTTCT'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store the sequence, defining a SQ class instance\n",
    "virus_sq = virus_n.get_sq()\n",
    "print(f'Sequence length: {len(virus_sq)} nucleotides')\n",
    "\n",
    "# The sequence is a bit too long, lets show the first 1000 characters\n",
    "virus_sq[0:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "vulnerable-chart",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:29.143918Z",
     "iopub.status.busy": "2021-09-08T13:53:29.128574Z",
     "iopub.status.idle": "2021-09-08T13:53:29.151464Z",
     "shell.execute_reply": "2021-09-08T13:53:29.152004Z"
    },
    "papermill": {
     "duration": 0.081295,
     "end_time": "2021-09-08T13:53:29.152179",
     "exception": false,
     "start_time": "2021-09-08T13:53:29.070884",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>                            <div id=\"d33f41da-1193-49f2-b253-36b130db5fab\" class=\"plotly-graph-div\" style=\"height:200px; width:100%;\"></div>            <script type=\"text/javascript\">                require([\"plotly\"], function(Plotly) {                    window.PLOTLYENV=window.PLOTLYENV || {};                                    if (document.getElementById(\"d33f41da-1193-49f2-b253-36b130db5fab\")) {                    Plotly.newPlot(                        \"d33f41da-1193-49f2-b253-36b130db5fab\",                        [{\"marker\": {\"color\": \"rgb(26, 118, 255)\"}, \"name\": \"SEQ1\", \"orientation\": \"h\", \"text\": [\"0.299\", \"0.321\", \"0.196\", \"0.184\"], \"textposition\": \"outside\", \"type\": \"bar\", \"x\": [0.299, 0.321, 0.196, 0.184], \"y\": [\"Adenine\", \"Thymine\", \"Guanine\", \"Cytosine\"]}],                        {\"font\": {\"family\": \"sans-serif\", \"size\": 12}, \"height\": 200, \"margin\": {\"b\": 10, \"l\": 40, \"r\": 40, \"t\": 50}, \"template\": {\"data\": {\"bar\": [{\"error_x\": {\"color\": \"#2a3f5f\"}, \"error_y\": {\"color\": \"#2a3f5f\"}, \"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"bar\"}], \"barpolar\": [{\"marker\": {\"line\": {\"color\": \"white\", \"width\": 0.5}}, \"type\": \"barpolar\"}], \"carpet\": [{\"aaxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"baxis\": {\"endlinecolor\": \"#2a3f5f\", \"gridcolor\": \"#C8D4E3\", \"linecolor\": \"#C8D4E3\", \"minorgridcolor\": \"#C8D4E3\", \"startlinecolor\": \"#2a3f5f\"}, \"type\": \"carpet\"}], \"choropleth\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"choropleth\"}], \"contour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"contour\"}], \"contourcarpet\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"contourcarpet\"}], \"heatmap\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmap\"}], \"heatmapgl\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"heatmapgl\"}], \"histogram\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"histogram\"}], \"histogram2d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2d\"}], \"histogram2dcontour\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"histogram2dcontour\"}], \"mesh3d\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"type\": \"mesh3d\"}], \"parcoords\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"parcoords\"}], \"pie\": [{\"automargin\": true, \"type\": \"pie\"}], \"scatter\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter\"}], \"scatter3d\": [{\"line\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatter3d\"}], \"scattercarpet\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattercarpet\"}], \"scattergeo\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergeo\"}], \"scattergl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattergl\"}], \"scattermapbox\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scattermapbox\"}], \"scatterpolar\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolar\"}], \"scatterpolargl\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterpolargl\"}], \"scatterternary\": [{\"marker\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"type\": \"scatterternary\"}], \"surface\": [{\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}, \"colorscale\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"type\": \"surface\"}], \"table\": [{\"cells\": {\"fill\": {\"color\": \"#EBF0F8\"}, \"line\": {\"color\": \"white\"}}, \"header\": {\"fill\": {\"color\": \"#C8D4E3\"}, \"line\": {\"color\": \"white\"}}, \"type\": \"table\"}]}, \"layout\": {\"annotationdefaults\": {\"arrowcolor\": \"#2a3f5f\", \"arrowhead\": 0, \"arrowwidth\": 1}, \"autotypenumbers\": \"strict\", \"coloraxis\": {\"colorbar\": {\"outlinewidth\": 0, \"ticks\": \"\"}}, \"colorscale\": {\"diverging\": [[0, \"#8e0152\"], [0.1, \"#c51b7d\"], [0.2, \"#de77ae\"], [0.3, \"#f1b6da\"], [0.4, \"#fde0ef\"], [0.5, \"#f7f7f7\"], [0.6, \"#e6f5d0\"], [0.7, \"#b8e186\"], [0.8, \"#7fbc41\"], [0.9, \"#4d9221\"], [1, \"#276419\"]], \"sequential\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]], \"sequentialminus\": [[0.0, \"#0d0887\"], [0.1111111111111111, \"#46039f\"], [0.2222222222222222, \"#7201a8\"], [0.3333333333333333, \"#9c179e\"], [0.4444444444444444, \"#bd3786\"], [0.5555555555555556, \"#d8576b\"], [0.6666666666666666, \"#ed7953\"], [0.7777777777777778, \"#fb9f3a\"], [0.8888888888888888, \"#fdca26\"], [1.0, \"#f0f921\"]]}, \"colorway\": [\"#636efa\", \"#EF553B\", \"#00cc96\", \"#ab63fa\", \"#FFA15A\", \"#19d3f3\", \"#FF6692\", \"#B6E880\", \"#FF97FF\", \"#FECB52\"], \"font\": {\"color\": \"#2a3f5f\"}, \"geo\": {\"bgcolor\": \"white\", \"lakecolor\": \"white\", \"landcolor\": \"white\", \"showlakes\": true, \"showland\": true, \"subunitcolor\": \"#C8D4E3\"}, \"hoverlabel\": {\"align\": \"left\"}, \"hovermode\": \"closest\", \"mapbox\": {\"style\": \"light\"}, \"paper_bgcolor\": \"white\", \"plot_bgcolor\": \"white\", \"polar\": {\"angularaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"radialaxis\": {\"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\"}}, \"scene\": {\"xaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"yaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}, \"zaxis\": {\"backgroundcolor\": \"white\", \"gridcolor\": \"#DFE8F3\", \"gridwidth\": 2, \"linecolor\": \"#EBF0F8\", \"showbackground\": true, \"ticks\": \"\", \"zerolinecolor\": \"#EBF0F8\"}}, \"shapedefaults\": {\"line\": {\"color\": \"#2a3f5f\"}}, \"ternary\": {\"aaxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"baxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}, \"bgcolor\": \"white\", \"caxis\": {\"gridcolor\": \"#DFE8F3\", \"linecolor\": \"#A2B1C6\", \"ticks\": \"\"}}, \"title\": {\"x\": 0.05}, \"xaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}, \"yaxis\": {\"automargin\": true, \"gridcolor\": \"#EBF0F8\", \"linecolor\": \"#EBF0F8\", \"ticks\": \"\", \"title\": {\"standoff\": 15}, \"zerolinecolor\": \"#EBF0F8\", \"zerolinewidth\": 2}}}, \"title\": {\"text\": \"<b>DNA SEQUENCE CONTENT</b>\"}},                        {\"responsive\": true}                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('d33f41da-1193-49f2-b253-36b130db5fab');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })                };                });            </script>        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC-Content Sequence 1: 37.97%\n"
     ]
    }
   ],
   "source": [
    "# The virus contains a heavy portion of Thymine & Adenine (63%)\n",
    "virus_sq.freq(fheight=200)\n",
    "print(f'GC-Content Sequence 1: {round(virus_sq.gc()*100,3)}%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "another-allocation",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:29.236171Z",
     "iopub.status.busy": "2021-09-08T13:53:29.231049Z",
     "iopub.status.idle": "2021-09-08T13:53:29.676499Z",
     "shell.execute_reply": "2021-09-08T13:53:29.675983Z"
    },
    "papermill": {
     "duration": 0.489154,
     "end_time": "2021-09-08T13:53:29.676648",
     "exception": false,
     "start_time": "2021-09-08T13:53:29.187494",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid Chains: 1208\n"
     ]
    }
   ],
   "source": [
    "# Get all amino acid chains above default = 1\n",
    "print(f'Amino Acid Chains: {len(virus_sq.ORF_protein())}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "taken-mambo",
   "metadata": {
    "papermill": {
     "duration": 0.037516,
     "end_time": "2021-09-08T13:53:29.750068",
     "exception": false,
     "start_time": "2021-09-08T13:53:29.712552",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "### <b><span style='color:#FF5733'>5.2</span> | FUNCTIONAL PROTEINS & OLIGOPEPTIDES</b>\n",
    "- To not complicate things too much, in <code>Section 3.6</code>, it wasn't mentioned that upon translation, our <b>amino acid chains</b> aren't actually all proteins.\n",
    "- There are some extra criteria these chains must meet in order to be classified as proteins. <b>Functional proteins</b> are chains above <b>20 amino acids</b>.\n",
    "- Smaller chains are called [oligopeptides](https://en.wikipedia.org/wiki/Oligopeptide) (2-20 amino acids) & have other functionalities.\n",
    "- Let's select the largest amino acid found in the genome & search for it in a databse, so we can identify it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "forbidden-croatia",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T13:53:29.837287Z",
     "iopub.status.busy": "2021-09-08T13:53:29.832158Z",
     "iopub.status.idle": "2021-09-08T13:53:30.740041Z",
     "shell.execute_reply": "2021-09-08T13:53:30.740600Z"
    },
    "papermill": {
     "duration": 0.953473,
     "end_time": "2021-09-08T13:53:30.740813",
     "exception": false,
     "start_time": "2021-09-08T13:53:29.787340",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid Chains: 243\n",
      "Largest Amino Acid: MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHLKDGTCGLVEVEKGVLPQLEQPYVFIKRSDARTAPHGHVMVELVAELEGIQYGRSGETLGVLVPHVGEIPVAYRKVLLRKNGNKGAGGHSYGADLKSFDLGDELGTDPYEDFQENWNTKHSSGVTRELMRELNGGAYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKRGVYCCREHEHEIAWYTERSEKSYELQTPFEIKLAKKFDTFNGECPNFVFPLNSIIKTIQPRVEKKKLDGFMGRIRSVYPVASPNECNQMCLSTLMKCDHCGETSWQTGDFVKATCEFCGTENLTKEGATTCGYLPQNAVVKIYCPACHNSEVGPEHSLAEYHNESGLKTILRKGGRTIAFGGCVFSYVGCHNKCAYWVPRASANIGCNHTGVVGEGSEGLNDNLLEILQKEKVNINIVGDFKLNEEIAIILASFSASTSAFVETVKGLDYKAFKQIVESCGNFKVTKGKAKKGAWNIGEQKSILSPLYAFASEAARVVRSIFSRTLETAQNSVRVLQKAAITILDGISQYSLRLIDAMMFTSDLATNNLVVMAYITGGVVQLTSQWLTNIFGTVYEKLKPVLDWLEEKFKEGVEFLRDGWEIVKFISTCACEIVGGQIVTCAKEIKESVQTFFKLVNKFLALCADSIIIGGAKLKALNLGETFVTHSKGLYRKCVKSREETGLLMPLKAPKEIIFLEGETLPTEVLTEEVVLKTGDLQPLEQPTSEAVEAPLVGTPVCINGLMLLEIKDTEKYCALAPNMMVTNNTFTLKGGAPTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVELGTEVNEFACVVADAVIKTLQPVSELLTPLGIDLDEWSMATYYLFDESGEFKLASHMYCSFYPPDEDEEEGDCEEEEFEPSTQYEYGTEDDYQGKPLEFGATSAALQPEEEQEEDWLDDDSQQTVGQQDGSEDNQTTTIQTIVEVQPQLEMELTPVVQTIEVNSFSGYLKLTDNVYIKNADIVEEAKKVKPTVVVNAANVYLKHGGGVAGALNKATNNAMQVESDDYIATNGPLKVGGSCVLSGHNLAKHCLHVVGPNVNKGEDIQLLKSAYENFNQHEVLLAPLLSAGIFGADPIHSLRVCVDTVRTNVYLAVFDKNLYDKLVSSFLEMKSEKQVEQKIAEIPKEEVKPFITESKPSVEQRKQDDKKIKACVEEVTTTLEETKFLTENLLLYIDINGNLHPDSATLVSDIDITFLKKDAPYIVGDVVQEGVLTAVVIPTKKAGGTTEMLAKALRKVPTDNYITTYPGQGLNGYTVEEAKTVLKKCKSAFYILPSIISNEKQEILGTVSWNLREMLAHAEETRKLMPVCVETKAIVSTIQRKYKGIKIQEGVVDYGARFYFYTSKTTVASLINTLNDLNETLVTMPLGYVTHGLNLEEAARYMRSLKVPATVSVSSPDAVTAYNGYLTSSSKTPEEHFIETISLAGSYKDWSYSGQSTQLGIEFLKRGDKSVYYTSNPTTFHLDGEVITFDNLKTLLSLREVRTIKVFTTVDNINLHTQVVDMSMTYGQQFGPTYLDGADVTKIKPHNSHEGKTFYVLPNDDTLRVEAFEYYHTTDPSFLGRYMSALNHTKKWKYPQVNGLTSIKWADNNCYLATALLTLQQIELKFNPPALQDAYYRARAGEAANFCALILAYCNKTVGELGDVRETMSYLFQHANLDSCKRVLNVVCKTCGQQQTTLKGVEAVMYMGTLSYEQFKKGVQIPCTCGKQATKYLVQQESPFVMMSAPPAQYELKHGTFTCASEYTGNYQCGHYKHITSKETLYCIDGALLTKSSEYKGPITDVFYKENSYTTTIKPVTYKLDGVVCTEIDPKLDNYYKKDNSYFTEQPIDLVPNQPYPNASFDNFKFVCDNIKFADDLNQLTGYKKPASRELKVTFFPDLNGDVVAIDYKHYTPSFKKGAKLLHKPIVWHVNNATNKATYKPNTWCIRCLWSTKPVETSNSFDVLKSEDAQGMDNLACEDLKPVSEEVVENPTIQKDVLECNVKTTEVVGDIILKPANNSLKITEEVGHTDLMAAYVDNSSLTIKKPNELSRVLGLKTLATHGLAAVNSVPWDTIANYAKPFLNKVVSTTTNIVTRCLNRVCTNYMPYFFTLLLQLCTFTRSTNSRIKASMPTTIAKNTVKSVGKFCLEASFNYLKSPNFSKLINIIIWFLLLSVCLGSLIYSTAALGVLMSNLGMPSYCTGYREGYLNSTNVTIATYCTGSIPCSVCLSGLDSLDTYPSLETIQITISSFKWDLTAFGLVAEWFLAYILFTRFFYVLGLAAIMQLFFSYFAVHFISNSWLMWLIINLVQMAPISAMVRMYIFFASFYYVWKSYVHVVDGCNSSTCMMCYKRNRATRVECTTIVNGVRRSFYVYANGGKGFCKLHNWNCVNCDTFCAGSTFISDEVARDLSLQFKRPINPTDQSSYIVDSVTVKNGSIHLYFDKAGQKTYERHSLSHFVNLDNLRANNTKGSLPINVIVFDGKSKCEESSAKSASVYYSQLMCQPILLLDQALVSDVGDSAEVAVKMFDAYVNTFSSTFNVPMEKLKTLVATAEAELAKNVSLDNVLSTFISAARQGFVDSDVETKDVVECLKLSHQSDIEVTGDSCNNYMLTYNKVENMTPRDLGACIDCSARHINAQVAKSHNIALIWNVKDFMSLSEQLRKQIRSAAKKNNLPFKLTCATTRQVVNVVTTKIALKGGKIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGYKAIDGGVTRDIASTDTCFANKHADFDTWFSQRGGSYTNDKACPLIAAVITREVGFVVPGLPGTILRTTNGDFLHFLPRVFSAVGNICYTPSKLIEYTDFATSACVLAAECTIFKDASGKPVPYCYDTNVLEGSVAYESLRPDTRYVLMDGSIIQFPNTYLEGSVRVVTTFDSEYCRHGTCERSEAGVCVSTSGRWVLNNDYYRSLPGVFCGVDAVNLLTNMFTPLIQPIGALDISASIVAGGIVAIVVTCLAYYFMRFRRAFGEYSHVVAFNTLLFLMSFTVLCLTPVYSFLPGVYSVIYLYLTFYLTNDVSFLAHIQWMVMFTPLVPFWITIAYIICISTKHFYWFFSNYLKRRVVFNGVSFSTFEEAALCTFLLNKEMYLKLRSDVLLPLTQYNRYLALYNKYKYFSGAMDTTSYREAACCHLAKALNDFSNSGSDVLYQPPQTSITSAVLQSGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTSEDMLNPNYEDLLIRKSNHNFLVQAGNVQLRVIGHSMQNCVLKLKVDTANPKTPKYKFVRIQPGQTFSVLACYNGSPSGVYQCAMRPNFTIKGSFLNGSCGSVGFNIDYDCVSFCYMHHMELPTGVHAGTDLEGNFYGPFVDRQTAQAAGTDTTITVNVLAWLYAAVINGDRWFLNRFTTTLNDFNLVAMKYNYEPLTQDHVDILGPLSAQTGIAVLDMCASLKELLQNGMNGRTILGSALLEDEFTPFDVVRQCSGVTFQSAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFAMGIIAMSAFAMMFVKHKHAFLCLFLLPSLATVAYFNMVYMPASWVMRIMTWLDMVDTSLSGFKLKDCVMYASAVVLLILMTARTVYDDGARRVWTLMNVLTLVYKVYYGNALDQAISMWALIISVTSNYSGVVTTVMFLARGIVFMCVEYCPIFFITGNTLQCIMLVYCFLGYFCTCYFGLFCLLNRYFRLTLGVYDYLVSTQEFRYMNSQGLLPPKNSIDAFKLNIKLLGVGGKPCIKVATVQSKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTTEAFEKMVSLLSVLLSMQGAVDINKLCEEMLDNRATLQAIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAKSEFDRDAAMQRKLEKMADQAMTQMYKQARSEDKRAKVTSAMQTMLFTMLRKLDNDALNNIINNARDGCVPLNIIPLTTAAKLMVVIPDYNTYKNTCDGTTFTYASALWEIQQVVDADSKIVQLSEISMDNSPNLAWPLIVTALRANSAVKLQNNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLSDLQDLKWARFPKSDGTGTIYTELEPPCRFVTDTPKGPKVKYLYFIKGLNNLNRGMVLGSLAATVRLQAGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLCTHTGTGQAITVTPEANMDQESFGGASCCLYCRCHIDHPNPKGFCDLKGKYVQIPTTCANDPVGFTLKNTVCTVCGMWKGYGCSCDQLREPMLQSADAQSFLNGFAV\n",
      "Length of Largest Amino Acid: 4405\n"
     ]
    }
   ],
   "source": [
    "# For convenience, let's select a subset above 50 amino acids\n",
    "print(f'Amino Acid Chains: {len(virus_sq.ORF_protein(mins=50))}')\n",
    "print(f'Largest Amino Acid: {virus_sq.ORF_protein(mins=50)[0]}')\n",
    "print(f'Length of Largest Amino Acid: {len(virus_sq.ORF_protein(mins=50)[0])}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "congressional-curve",
   "metadata": {
    "papermill": {
     "duration": 0.036749,
     "end_time": "2021-09-08T13:53:30.816506",
     "exception": false,
     "start_time": "2021-09-08T13:53:30.779757",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "#### <b><span style='color:#FF5733'>SEARCHING DATABASES</span></b>\n",
    "\n",
    "- Next, we can use this string; copy & paste this protein & search for it using the [PSI BLAST](https://www.ebi.ac.uk/Tools/sss/psiblast/) in order to find proteins already in a predefined database. \n",
    "- For our current protein, which we wish to identify, the results indicate that a 100% match was found with an already existing sequence in the database: [Replicase polyprotein 1a](https://pubchem.ncbi.nlm.nih.gov/protein/P0C6U8) as shown by a [temporary job link](https://www.ebi.ac.uk/Tools/services/web/toolresult.ebi?jobId=psiblast-I20210801-145740-0891-84805416-p1m)\n",
    "- Through database searches & the <b>local sequence alignment</b> algorithm it uses, we can get a more clear picture of what functions the protein of interest can have. \n",
    "- Searches are stored for 7 days, after which a new search needs to be made when the link expires."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "israeli-workstation",
   "metadata": {
    "papermill": {
     "duration": 0.036501,
     "end_time": "2021-09-08T13:53:30.891438",
     "exception": false,
     "start_time": "2021-09-08T13:53:30.854937",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "<div style=\"color:white;\n",
    "       display:fill;\n",
    "       border-radius:5px;\n",
    "       background-color:#FF5733;\n",
    "       font-size:220%;\n",
    "       font-family:Nexa;\n",
    "       letter-spacing:0.5px\">\n",
    "    <p style=\"padding: 20px;\n",
    "          color:white;\">\n",
    "        <b>6 |</b> SUMMARY\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "#### <b><span style='color:#FF5733'>CUSTOM CLASSES</span></b>\n",
    "- In this notebook, we have defined a class that stores <b>basic information</b> about a specific sequence of interest. \n",
    "- Another class was also defined in order to read a commonly encountered format type <b>FASTA</b>. \n",
    "\n",
    "#### <b><span style='color:#FF5733'>OVERVIEW</span></b>\n",
    "- We also looked at common sequence related operations such as <b>transcription</b> & <b>translation</b>, which are quite critical concepts.\n",
    "- We also looked at some basic visualisation of the nucleotide & amino acid distributions, which can be helpful when comparing not only the <b>count</b>, but also the percentage of each alphabet for the corresponding sequence type.\n",
    "\n",
    "#### <b><span style='color:#FF5733'>WHERE TO FROM HERE</span></b>\n",
    "- As with the <b>BioPython</b> module, we have defined classes which store sequence data & which can be expanded further.\n",
    "- The class, <b>SQ</b> will be used as a basis for other sequence related operations, such as <b>pairwise</b> & <b>multiple sequence alignment</b>, which due to their content size alone, can have their own dedicated classes, similar to the class <b>READ_SEQ</b>, which is dedicated to reading and passing on the sequence information to class <b>SQ</b>.\n",
    "- The notebook will of course be continuously updated and futher functions will be added in the future."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 15.946061,
   "end_time": "2021-09-08T13:53:31.738306",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2021-09-08T13:53:15.792245",
   "version": "2.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
